File: apps/patter-voice-sdk/libraries/python/getpatter/llm/openai.py
"""OpenAI LLM for Patter pipeline mode."""

from __future__ import annotations

import os
from typing import ClassVar

from getpatter.services.llm_loop import OpenAILLMProvider as _OpenAILLM

__all__ = ["LLM"]


class LLM(_OpenAILLM):
    """OpenAI Chat Completions LLM provider.

    Example::

        from getpatter.llm import openai

        llm = openai.LLM()                            # reads OPENAI_API_KEY
        llm = openai.LLM(api_key="sk-...", model="gpt-4o-mini")
    """

    provider_key: ClassVar[str] = "openai"

    def __init__(
        self,
        api_key: str | None = None,
        *,
        model: str = "gpt-4o-mini",
        **kwargs,
    ) -> None:
        key = api_key or os.environ.get("OPENAI_API_KEY")
        if not key:
            raise ValueError(
                "OpenAI LLM requires an api_key. Pass api_key='sk-...' or "
                "set OPENAI_API_KEY in the environment."
            )
        super().__init__(api_key=key, model=model, **kwargs)

File: apps/patter-voice-sdk/libraries/typescript/src/llm/openai.ts
/** OpenAI LLM for Patter pipeline mode. */
import { OpenAILLMProvider as _OpenAILLM } from "../llm-loop";

/** Constructor options for the OpenAI Chat Completions `LLM` adapter. */
export interface OpenAILLMOptions {
  /** API key. Falls back to OPENAI_API_KEY env var when omitted. */
  apiKey?: string;
  /** Chat Completions model id. Defaults to ``"gpt-4o-mini"``. */
  model?: string;
  /** Sampling temperature [0, 2]. */
  temperature?: number;
  /** Max tokens in the assistant response (sent as ``max_completion_tokens``). */
  maxTokens?: number;
  /** OpenAI-style ``response_format`` for JSON mode / structured outputs. */
  responseFormat?: Record<string, unknown>;
  /** Whether to allow parallel tool calls. */
  parallelToolCalls?: boolean;
  /** ``"auto" | "none" | "required"`` or a specific tool object. */
  toolChoice?: string | Record<string, unknown>;
  /** Sampling seed for reproducible outputs. */
  seed?: number;
  /** Nucleus sampling cutoff in [0, 1]. */
  topP?: number;
  /** Penalty in [-2, 2] applied to repeated tokens. */
  frequencyPenalty?: number;
  /** Penalty in [-2, 2] applied to seen tokens. */
  presencePenalty?: number;
  /** Stop sequence(s). */
  stop?: string | string[];
}

/**
 * OpenAI Chat Completions LLM provider.
 *
 * @example
 * ```ts
 * import * as openai from "getpatter/llm/openai";
 * const llm = new openai.LLM();                           // reads OPENAI_API_KEY
 * const llm = new openai.LLM({ apiKey: "sk-...", model: "gpt-4o-mini", temperature: 0.4 });
 * ```
 */
export class LLM extends _OpenAILLM {
  static readonly providerKey = "openai";
  constructor(opts: OpenAILLMOptions = {}) {
    const key = opts.apiKey ?? process.env.OPENAI_API_KEY;
    if (!key) {
      throw new Error(
        "OpenAI LLM requires an apiKey. Pass { apiKey: 'sk-...' } or set OPENAI_API_KEY.",
      );
    }
    super(key, opts.model ?? "gpt-4o-mini", {
      temperature: opts.temperature,
      maxTokens: opts.maxTokens,
      responseFormat: opts.responseFormat,
      parallelToolCalls: opts.parallelToolCalls,
      toolChoice: opts.toolChoice,
      seed: opts.seed,
      topP: opts.topP,
      frequencyPenalty: opts.frequencyPenalty,
      presencePenalty: opts.presencePenalty,
      stop: opts.stop,
    });
  }
}
